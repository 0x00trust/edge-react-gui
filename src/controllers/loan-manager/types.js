// @flow

import { type BorrowEngine, type BorrowPlugin } from '../../plugins/borrow-plugins/types'
import { type LoanProgramEdge } from './store'

export type LoanAccount = {
  id: string,
  borrowPlugin: BorrowPlugin,
  borrowEngine: BorrowEngine,

  // Action Program Relationships
  programEdges: LoanProgramEdge[]
}

export type LoanAccountMap = {
  [walletId: string]: LoanAccount
}

export type LoanManager = {
  loanAccounts: LoanAccount[]
}
