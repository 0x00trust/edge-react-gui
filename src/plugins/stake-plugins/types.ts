// -----------------------------------------------------------------------------
// Stake Policy
// -----------------------------------------------------------------------------

import { EdgeCurrencyWallet } from 'edge-core-js'

export interface AssetId {
  pluginId: string
  currencyCode: string
}

// Defines what to display so that the user can identify the service provider
// of the staking policy.
export interface StakeProviderInfo {
  // Card subtitle
  displayName: string

  // CDN icon
  pluginId: string
  stakeProviderId: string
}

export interface StakePolicy {
  // Internal policy id, unique across all stake policies offered by Edge
  stakePolicyId: string

  // Some unique grouping information to display in the policy card to
  // differentiate between stake policies that share the same input stakeAssets
  stakeProviderInfo?: StakeProviderInfo

  // A percentage number representing the yield per year
  apy: number

  // The assets which can be earned
  rewardAssets: AssetId[]

  // The assets which must be staked
  stakeAssets: AssetId[]
}

// -----------------------------------------------------------------------------
// Change Quote
// -----------------------------------------------------------------------------
export interface ChangeQuoteRequest {
  action: 'stake' | 'unstake' | 'claim'
  stakePolicyId: string
  currencyCode: string
  nativeAmount: string
  wallet: EdgeCurrencyWallet
}

export interface QuoteAllocation {
  allocationType: 'stake' | 'unstake' | 'claim' | 'fee'
  pluginId: string
  currencyCode: string
  nativeAmount: string
}

export interface ChangeQuote {
  allocations: QuoteAllocation[]
  approve: () => Promise<void>
}

// -----------------------------------------------------------------------------
// Stake Position
// -----------------------------------------------------------------------------

export interface StakePositionRequest {
  stakePolicyId: string
  wallet: EdgeCurrencyWallet
}

export interface PositionAllocation {
  // The of asset for this allocation
  pluginId: string
  currencyCode: string
  // The of the allocation
  allocationType: 'staked' | 'unstaked' | 'earned'
  // Amount of the asset allocated
  nativeAmount: string
  /*
  A date/time when the allocation is available.
  Example: earned allocations with a future date are not available,
  but earned allocations with a past date are available to be earned.
  For some tokens (e.g. FIO), there is no earned allocation; rather there is
  an unstaked allocation which is locked until the date.
  */
  locktime?: Date
}

export interface StakePosition {
  allocations: PositionAllocation[]
  canStake: boolean
  canUnstake: boolean
  canClaim: boolean
}

// -----------------------------------------------------------------------------
// Stake Plugin
// -----------------------------------------------------------------------------

export interface StakePlugin {
  policies: StakePolicy[]
  fetchChangeQuote: (request: ChangeQuoteRequest) => Promise<ChangeQuote>
  fetchStakePosition: (request: StakePositionRequest) => Promise<StakePosition>
}
