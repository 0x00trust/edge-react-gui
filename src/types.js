/**
 * Created by paul on 8/16/17.
 */

// @flow

export interface ABCDenomination {
  name:string,
  multiplier:string,
  symbol:string|null
}

export interface ABCMetaToken {
  currencyCode:string,
  currencyName:string,
  denominations:Array<ABCDenomination>,
  contractAddress:string|null,
  symbolImage:string|null
}

export class GUIWallet {
  id:string
  type:string
  name:string
  primaryNativeBalance:string
  nativeBalances: { [currencyCode: string]: string }
  currencyNames: { [currencyCode: string]: string }
  currencyCode:string
  isoFiatCurrencyCode:string
  fiatCurrencyCode:string
  denominations:Array<ABCDenomination>
  allDenominations: { [currencyCode: string]: { [denomination: string]: ABCDenomination } }
  symbolImage:string
  metaTokens:Array<ABCMetaToken>
  sortIndex:number
  archived:boolean
  deleted:boolean
  constructor (
    id:string,
    type:string,
    name:string,
    primaryNativeBalance:string,
    nativeBalances: { [currencyCode: string]: string },
    currencyNames: { [currencyCode: string]: string },
    currencyCode:string,
    isoFiatCurrencyCode:string,
    fiatCurrencyCode:string,
    denominations:Array<ABCDenomination>,
    allDenominations: { [currencyCode: string]: { [denomination: string]: ABCDenomination } },
    symbolImage:string,
    metaTokens:Array<ABCMetaToken>,
    sortIndex:number,
    archived:boolean,
    deleted:boolean
  ) {
    this.id = id
    this.type = type
    this.name = name
    this.primaryNativeBalance = primaryNativeBalance
    this.nativeBalances = nativeBalances
    this.currencyNames = currencyNames
    this.currencyCode = currencyCode
    this.isoFiatCurrencyCode = isoFiatCurrencyCode
    this.fiatCurrencyCode = fiatCurrencyCode
    this.denominations = denominations
    this.allDenominations = allDenominations
    this.symbolImage = symbolImage
    this.metaTokens = metaTokens
    this.sortIndex = sortIndex
    this.archived = archived
    this.deleted = deleted
  }
}
