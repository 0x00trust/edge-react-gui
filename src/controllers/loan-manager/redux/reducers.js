// @flow

import { type Reducer, combineReducers } from 'redux'

import { type Action } from '../../../types/reduxTypes'
import { type LoanAccountMap } from '../types'

export type LoanManagerState = {
  +loanAccounts: LoanAccountMap
}

export const loanManager: Reducer<LoanManagerState, Action> = combineReducers({
  loanAccounts(state: LoanAccountMap = {}, action: Action): LoanAccountMap {
    switch (action.type) {
      case 'LOAN_MANAGER/SET_LOAN_ACCOUNT': {
        const { loanAccount } = action
        return {
          ...state,
          [loanAccount.id]: loanAccount
        }
      }
      case 'LOAN_MANAGER/DELETE_LOAN_ACCOUNT': {
        const { id } = action
        const { [id]: _, ...remainingState } = state
        return remainingState
      }
      case 'LOGOUT': {
        return {}
      }
      default:
        return state
    }
  }
})
