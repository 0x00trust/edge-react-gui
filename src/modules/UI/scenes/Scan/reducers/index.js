// @flow

import { addressModalVisible } from './addressModalVisible.js'
import { parsedUri } from './parsedUri.js'
import { scanEnabled } from './scanEnabled.js'
import { torchEnabled } from './torchEnabled.js'

export { addressModalVisible, scanEnabled, torchEnabled, parsedUri }
