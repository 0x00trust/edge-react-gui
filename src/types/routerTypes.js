// @flow

import { type EdgeCurrencyInfo, type EdgeCurrencyWallet, type EdgeTransaction, type JsonObject, type OtpError } from 'edge-core-js'
import * as React from 'react'
import * as Flux from 'react-native-router-flux'

import type { ExchangedFlipInputAmounts } from '../components/themed/ExchangedFlipInput.js'
import { type BorrowEngine, type BorrowPlugin } from '../plugins/borrow-plugins/types'
import { type FiatPluginEnterAmountResponse, type FiatPluginGetMethodsResponse } from '../plugins/gui/fiatPluginTypes.js'
import { type ChangeQuoteRequest, type StakePolicy, type StakePosition } from '../plugins/stake-plugins'
import { type GuiPlugin } from './GuiPluginTypes.js'
import {
  type CreateWalletType,
  type FeeOption,
  type FioConnectionWalletItem,
  type FioDomain,
  type FioRequest,
  type GuiFiatType,
  type GuiMakeSpendInfo,
  type GuiSwapInfo,
  type WcConnectionInfo
} from './types.js'
import { type UriQueryMap } from './WebTypes'

type PluginViewParams = {|
  // The GUI plugin we are showing the user:
  plugin: GuiPlugin,

  // Set these to add stuff to the plugin URI:
  deepPath?: string,
  deepQuery?: UriQueryMap
|}
/**
 * Defines the acceptable route parameters for each scene key.
 */
export type ParamList = {
  // Top-level router:
  root: {||},
  login: {||},
  edge: {||},

  // Logged-in scenes:
  changeMiningFee: {|
    guiMakeSpendInfo: GuiMakeSpendInfo,
    maxSpendSet: boolean,
    onSubmit: (networkFeeOption: FeeOption, customNetworkFee: JsonObject) => void,
    wallet: EdgeCurrencyWallet
  |},
  changePassword: {||},
  changePin: {||},
  controlPanel: {||},
  createWalletAccountSelect: {|
    accountName: string,
    existingWalletId?: string,
    selectedFiat: GuiFiatType,
    selectedWalletType: CreateWalletType
  |},
  createWalletAccountSetup: {|
    accountHandle?: string,
    existingWalletId?: string,
    isReactivation?: boolean,
    selectedFiat: GuiFiatType,
    selectedWalletType: CreateWalletType
  |},
  createWalletChoice: {|
    selectedWalletType: CreateWalletType
  |},
  createWalletImport: {|
    selectedWalletType: CreateWalletType
  |},
  createWalletName: {|
    cleanedPrivateKey?: string,
    selectedFiat: GuiFiatType,
    selectedWalletType: CreateWalletType
  |},
  createWalletReview: {|
    cleanedPrivateKey?: string, // for creating wallet from import private key
    selectedFiat: GuiFiatType,
    selectedWalletType: CreateWalletType,
    walletName: string
  |},
  createWalletSelectCrypto: {||},
  createWalletSelectFiat: {|
    selectedWalletType: CreateWalletType,
    cleanedPrivateKey?: string
  |},
  currencyNotificationSettings: {|
    currencyInfo: EdgeCurrencyInfo
  |},
  currencySettings: {|
    currencyInfo: EdgeCurrencyInfo
  |},
  defaultFiatSetting: {||},
  edgeLogin: {||},
  editToken: {|
    currencyCode?: string,
    displayName?: string,
    multiplier?: string,
    networkLocation?: JsonObject,
    tokenId?: string, // Acts like "add token" if this is missing
    walletId: string
  |},
  exchange: {||},
  exchangeQuote: {|
    swapInfo: GuiSwapInfo,
    onApprove: () => void
  |},
  exchangeQuoteProcessing: {||},
  exchangeScene: {||},
  exchangeSettings: {||},
  exchangeSuccess: {||},
  fioAddressDetails: {|
    fioAddressName: string,
    bundledTxs: number
  |},
  fioAddressList: {||},
  fioAddressRegister: {||},
  fioAddressRegisterSelectWallet: {|
    fioAddress: string,
    selectedWallet: EdgeCurrencyWallet,
    selectedDomain: FioDomain,
    isFallback?: boolean
  |},
  fioAddressRegisterSuccess: {|
    fioName: string,
    expiration?: string
  |},
  fioAddressSettings: {|
    fioWallet: EdgeCurrencyWallet,
    fioAddressName: string,
    bundledTxs?: number,
    showAddBundledTxs?: boolean,
    refreshAfterAddBundledTxs?: boolean
  |},
  fioConnectToWalletsConfirm: {|
    fioWallet: EdgeCurrencyWallet,
    fioAddressName: string,
    walletsToConnect: FioConnectionWalletItem[],
    walletsToDisconnect: FioConnectionWalletItem[]
  |},
  fioDomainConfirm: {|
    fioName: string,
    paymentWallet: EdgeCurrencyWallet,
    fee: number,
    ownerPublicKey: string
  |},
  fioDomainRegister: {||},
  fioDomainRegisterSelectWallet: {|
    fioDomain: string,
    selectedWallet: EdgeCurrencyWallet
  |},
  fioDomainSettings: {|
    fioWallet: EdgeCurrencyWallet,
    fioDomainName: string,
    isPublic: boolean,
    expiration: string,
    showRenew?: boolean
  |},
  fioNameConfirm: {|
    fioName: string,
    paymentWallet: EdgeCurrencyWallet,
    fee: number,
    ownerPublicKey: string
  |},
  fioRequestConfirmation: {|
    amounts: ExchangedFlipInputAmounts
  |},
  fioRequestList: {||},
  fioRequestApproved: {|
    edgeTransaction: EdgeTransaction,
    thumbnailPath?: string
  |},
  fioSentRequestDetails: {|
    selectedFioSentRequest: FioRequest
  |},
  fioStakingChange: {
    change: 'add' | 'remove',
    currencyCode: string,
    walletId: string
  },
  fioStakingOverview: {
    currencyCode: string,
    walletId: string
  },
  guiPluginEnterAmount: {|
    headerTitle: string,
    onSubmit: (response: FiatPluginEnterAmountResponse) => Promise<void>,
    label1: string,
    label2: string,
    onChangeText: (fieldNum: number, value: string) => Promise<void>,
    convertValue: (sourceFieldNum: number, value: string) => Promise<string | void>,
    getMethods?: (methods: FiatPluginGetMethodsResponse) => void,
    initialAmount1?: string,
    headerIconUri?: string
  |},
  loanDashboard: {||},
  loanDetails: {|
    loanAccountId: string
  |},
  loanCreate: {|
    borrowEngine: BorrowEngine,
    borrowPlugin: BorrowPlugin
  |},
  loanCreateConfirmation: {|
    borrowEngine: BorrowEngine,
    borrowPlugin: BorrowPlugin,
    destTokenId: string,
    destWallet: EdgeCurrencyWallet,
    isDestBank: boolean,
    nativeDestAmount: string,
    nativeSrcAmount: string,
    srcTokenId?: string,
    srcWallet: EdgeCurrencyWallet
  |},
  loanClose: {|
    loanAccountId: string
  |},
  loanAddCollateralScene: {|
    loanAccountId: string
  |},
  loanBorrowMoreScene: {|
    loanAccountId: string
  |},
  loanWithdrawCollateralScene: {|
    loanAccountId: string
  |},
  loanRepayScene: {|
    loanAccountId: string
  |},
  loanStatus: {|
    actionQueueId: string
  |},
  loanCreateStatus: {|
    actionQueueId: string
  |},
  loanDetailsStatus: {|
    actionQueueId: string
  |},
  manageTokens: {|
    walletId: string
  |},
  notificationSettings: {||},
  otpRepair: {|
    otpError: OtpError
  |},
  otpSetup: {||},
  passwordRecovery: {||},
  pluginListBuy: {| direction: 'buy' |},
  pluginListSell: {| direction: 'sell' |},
  pluginViewBuy: PluginViewParams,
  pluginViewSell: PluginViewParams,
  pluginView: PluginViewParams,
  promotionSettings: {||},
  request: {||},
  scan: {|
    data?: 'sweepPrivateKey' | 'loginQR'
  |}, // TODO
  securityAlerts: {||},
  send: {|
    allowedCurrencyCodes?: string[],
    guiMakeSpendInfo?: GuiMakeSpendInfo,
    selectedWalletId?: string,
    selectedCurrencyCode?: string,
    isCameraOpen?: boolean,
    lockTilesMap?: {
      address?: boolean,
      wallet?: boolean,
      amount?: boolean
    },
    hiddenTilesMap?: {
      address?: boolean,
      amount?: boolean,
      fioAddressSelect?: boolean
    },
    infoTiles?: Array<{ label: string, value: string }>
  |},
  settingsOverview: {||},
  settingsOverviewTab: {||},
  spendingLimits: {||},
  stakeModify: {|
    walletId: string,
    stakePolicy: StakePolicy,
    stakePosition: StakePosition,
    modification: $PropertyType<ChangeQuoteRequest, 'action'>
  |},
  stakeClaim: { walletId: string, stakePolicy: StakePolicy },
  stakeOptions: { walletId: string, currencyCode: string },
  stakeOverview: { walletId: string, stakePolicy: StakePolicy },
  termsOfService: {||},
  testScene: {||},
  transactionDetails: {|
    edgeTransaction: EdgeTransaction,
    thumbnailPath?: string
  |},
  transactionList: {||},
  transactionsExport: {|
    sourceWallet: EdgeCurrencyWallet,
    currencyCode: string
  |},
  walletList: {||},
  walletListScene: {||},
  wcConnections: {||},
  wcDisconnect: {| wcConnectionInfo: WcConnectionInfo |},
  wcConnect: {|
    uri: string
  |}
}

/**
 * The global `Actions` object for navigation.
 */
export const Actions = {
  get currentParams(): any {
    // $FlowFixMe
    return Flux.Actions.currentParams
  },
  get currentScene(): $Keys<ParamList> {
    // $FlowFixMe
    return Flux.Actions.currentScene
  },

  drawerClose() {
    // $FlowFixMe
    Flux.Actions.drawerClose()
  },
  drawerOpen() {
    // $FlowFixMe
    Flux.Actions.drawerOpen()
  },

  jump<Name extends $Keys<ParamList>>(name: Name, params: $ElementType<ParamList, Name>): void {
    // $FlowFixMe
    Flux.Actions.jump(name, { route: { name, params } })
  },
  push<Name extends $Keys<ParamList>>(name: Name, params: $ElementType<ParamList, Name>): void {
    // $FlowFixMe
    Flux.Actions.push(name, { route: { name, params } })
  },
  replace<Name extends $Keys<ParamList>>(name: Name, params: $ElementType<ParamList, Name>): void {
    // $FlowFixMe
    Flux.Actions.replace(name, { route: { name, params } })
  },

  refresh(params: any): void {
    // $FlowFixMe
    Flux.Actions.refresh({ route: { name: Flux.Actions.currentScene, params } })
  },

  pop(): void {
    // $FlowFixMe
    Flux.Actions.pop()
  },
  popTo(name: $Keys<ParamList>): void {
    // $FlowFixMe
    Flux.Actions.popTo(name)
  }
}

type NavigationEvent = 'didBlur' | 'didFocus' | 'willBlur' | 'willFocus'

/**
 * The type of the `navigation` prop passed to each scene.
 */
export type NavigationProp<Name extends $Keys<ParamList>> = {
  // Whether this scene is in the foreground:
  addListener: (event: NavigationEvent, callback: () => void) => () => void,
  isFocused: () => boolean,

  // Going places:
  navigate: <Name extends $Keys<ParamList>>(name: Name, params: $ElementType<ParamList, Name>) => void,
  push: <Name extends $Keys<ParamList>>(name: Name, params: $ElementType<ParamList, Name>) => void,
  replace: <Name extends $Keys<ParamList>>(name: Name, params: $ElementType<ParamList, Name>) => void,
  setParams: (params: $ElementType<ParamList, Name>) => void,

  // Returning:
  goBack: () => void,
  pop: () => void,
  popToTop: () => void,

  // Drawer:
  closeDrawer: () => void,
  openDrawer: () => void,
  toggleDrawer: () => void,

  // Internals nobody should need to touch:
  state: mixed
}

/**
 * The type of the `route` prop passed to each scene.
 */
export type RouteProp<Name extends $Keys<ParamList>> = {
  name: Name,
  params: $ElementType<ParamList, Name>
}

/**
 * Adjusts the navigation prop to match the type definitions above.
 */
export function withNavigation<Props>(Component: React.ComponentType<Props>): React.StatelessFunctionalComponent<Props> {
  function WithNavigation(props: any) {
    const navigation: NavigationProp<'edge'> = {
      addListener(event, callback) {
        const remover = props.navigation.addListener(event, callback)
        return () => remover.remove()
      },
      isFocused() {
        return props.navigation.isFocused()
      },

      navigate(name, params) {
        props.navigation.navigate(name, { route: { name, params } })
      },
      push(name, params) {
        props.navigation.push(name, { route: { name, params } })
      },
      replace(...a) {
        // prev, next, params) {
        props.navigation.replace(...a) // prev, next, next, { route: { params } })
      },
      setParams(params) {
        props.navigation.setParams({ route: { name: Actions.currentScene, params } })
      },

      goBack() {
        props.navigation.goBack()
      },
      pop() {
        props.navigation.pop()
      },
      popToTop() {
        props.navigation.popToTop()
      },

      closeDrawer() {
        props.navigation.closeDrawer()
      },
      openDrawer() {
        props.navigation.openDrawer()
      },
      toggleDrawer() {
        props.navigation.toggleDrawer()
      },

      get state() {
        return props.navigation.state
      }
    }

    return <Component {...props} navigation={navigation} />
  }
  const displayName = Component.displayName ?? Component.name ?? 'Component'
  WithNavigation.displayName = `WithNavigation(${displayName})`
  return WithNavigation
}
