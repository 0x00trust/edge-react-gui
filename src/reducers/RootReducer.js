// @flow

import { type Reducer, combineReducers } from 'redux'

import { type ExchangeRatesState, exchangeRates } from '../modules/ExchangeRates/reducer.js'
import { type UiState, ui } from '../modules/UI/reducer.js'
import { type Action } from '../types/reduxTypes.js'
import { type ContactsState, contacts } from './ContactsReducer.js'
import { type CoreState, core } from './CoreReducer.js'
import { type CryptoExchangeState, cryptoExchange } from './CryptoExchangeReducer.js'
import { type PermissionsState, permissions } from './PermissionsReducer.js'

export type RootState = {
  // Next username to auto-fill at the login screen, or blank if none:
  +nextUsername: string | null,

  // Nested reducers:
  +contacts: ContactsState,
  +core: CoreState,
  +cryptoExchange: CryptoExchangeState,
  +exchangeRates: ExchangeRatesState,
  +permissions: PermissionsState,
  +ui: UiState
}

export const rootReducer: Reducer<RootState, Action> = combineReducers({
  nextUsername (state: string | null = null, action: Action): string | null {
    switch (action.type) {
      case 'DEEP_LINK_RECEIVED':
        return null
      case 'LOGOUT': {
        const { username = null } = action.data
        return username
      }
    }
    return state
  },

  // Nested reducers:
  contacts,
  core,
  cryptoExchange,
  exchangeRates,
  permissions,
  ui
})
