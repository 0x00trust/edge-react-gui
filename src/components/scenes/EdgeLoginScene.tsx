import { EdgeLobby } from 'edge-core-js'
import * as React from 'react'
import { ActivityIndicator, Image, View } from 'react-native'
import { sprintf } from 'sprintf-js'

import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import s from '../../locales/strings'
import { config } from '../../theme/appConfig'
import { useSelector } from '../../types/reactRedux'
import { NavigationBase, RouteProp } from '../../types/routerTypes'
import { WarningCard } from '../cards/WarningCard'
import { SceneWrapper } from '../common/SceneWrapper'
import { ButtonsModal } from '../modals/ButtonsModal'
import { Airship, showError } from '../services/AirshipInstance'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { TitleText } from '../text/TitleText'
import { Fade } from '../themed/Fade'
import { MainButton } from '../themed/MainButton'

interface Props {
  navigation: NavigationBase
  route: RouteProp<'edgeLogin'>
}

export const EdgeLoginScene = (props: Props) => {
  const { navigation, route } = props
  const { account } = useSelector(state => state.core)
  const { lobbyId } = route.params
  const [lobby, setLobby] = React.useState<EdgeLobby | undefined>(undefined)
  const [isApproved, setIsApproved] = React.useState<boolean>(false)
  const isLobby = lobby != null
  const isLoading = !isLobby || isApproved
  const warningMessage = sprintf(s.strings.edge_description_warning, config.appName)
  const theme = useTheme()
  const styles = getStyles(theme)

  useAsyncEffect(async () => {
    await account
      .fetchLobby(lobbyId)
      .then((lobby: EdgeLobby) => {
        setLobby(lobby)
      })
      .catch((e: any) => {
        if (e.message.includes('Account does not')) {
          showOkModal(s.strings.edge_login_failed, s.strings.edge_login_fail_stale_qr)
        } else {
          showError(e)
        }
        navigation.pop()
      })
  }, [lobbyId])

  const handleAccept = async () => {
    if (lobby?.loginRequest == null) return
    setIsApproved(true)
    const { loginRequest } = lobby
    await loginRequest
      .approve()
      .then(() => {
        navigation.pop()
        showOkModal(s.strings.send_scan_edge_login_success_title, s.strings.send_scan_edge_login_success_message)
      })
      .catch((e: any) => {
        navigation.pop()
        if (e.message.includes('Could not reach')) {
          showOkModal(s.strings.edge_login_failed, s.strings.edge_login_fail_message)
        } else {
          showError(e)
        }
      })
  }

  const showOkModal = async (title: string, message: string) => {
    return Airship.show<'ok' | undefined>(bridge => (
      <ButtonsModal bridge={bridge} buttons={{ ok: { label: s.strings.string_ok } }} message={message + '\n'} title={title} />
    ))
  }

  const handleDecline = () => navigation.goBack()

  return (
    <SceneWrapper background="theme">
      <View style={styles.container}>
        <View style={styles.header}>
          <Image style={styles.logo} source={theme.primaryLogo} resizeMode="contain" />
          <TitleText style={styles.appName}>{config.appName}</TitleText>
        </View>
        <View style={styles.loader}>
          <Fade visible={isLoading} noFadeIn>
            <ActivityIndicator color={theme.iconTappable} size="large" />
          </Fade>
        </View>
        <View style={styles.container}>
          <Fade visible={!isLoading} delay={125}>
            <WarningCard title={s.strings.string_warning} header={warningMessage} />
          </Fade>
          <View>
            <Fade visible={isLobby} delay={250}>
              <MainButton label={s.strings.accept_button_text} onPress={handleAccept} marginRem={1} />
            </Fade>
            <Fade visible={isLobby} delay={500}>
              <MainButton label={s.strings.string_cancel_cap} onPress={handleDecline} type="escape" marginRem={[0, 0, 1, 0]} />
            </Fade>
          </View>
        </View>
      </View>
    </SceneWrapper>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    alignSelf: 'center',
    justifyContent: 'space-between',
    flex: 1
  },
  logo: {
    height: theme.rem(6),
    width: theme.rem(12)
  },
  appName: {
    fontSize: theme.rem(1.25),
    paddingTop: theme.rem(1)
  },
  header: {
    alignItems: 'center'
  },
  loader: {
    opacity: 0.5,
    paddingVertical: theme.rem(1.5),
    height: theme.rem(4)
  }
}))
