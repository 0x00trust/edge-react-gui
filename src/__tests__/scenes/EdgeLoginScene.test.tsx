import { describe, expect, it } from '@jest/globals'
import { EdgeAccount } from 'edge-core-js'
import * as React from 'react'
import { Provider } from 'react-redux'
import { createRenderer } from 'react-test-renderer/shallow'
import { createStore } from 'redux'

import { EdgeLoginScene } from '../../components/scenes/EdgeLoginScene'
import { rootReducer } from '../../reducers/RootReducer'
import { RouteProp } from '../../types/routerTypes'
import { fakeNavigation } from '../../util/fake/fakeNavigation'
import { fakeNonce } from '../../util/fake/fakeNonce'
import { fakeRootState } from '../../util/fake/fakeRootState'

let account: EdgeAccount | undefined

describe('EdgeLoginScene', () => {
  const nonce = fakeNonce(0)
  it('should render with loading props', () => {
    const renderer = createRenderer()
    const route: RouteProp<'edgeLogin'> = {
      key: `edgeLogin-${nonce()}`,
      name: 'edgeLogin',
      params: {
        lobbyId: 'AmNsSBDVeF2837'
      }
    }
    const rootState: any = fakeRootState
    rootState.core = {
      account
    }
    const store = createStore(rootReducer, rootState)
    const actual = renderer.render(
      <Provider store={store}>
        <EdgeLoginScene route={route} navigation={fakeNavigation} />
      </Provider>
    )

    expect(actual).toMatchSnapshot()
  })
})
