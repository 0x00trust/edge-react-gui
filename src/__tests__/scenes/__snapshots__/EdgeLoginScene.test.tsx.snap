// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EdgeLoginScene should render with loading props 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <EdgeLoginScene
    navigation={
      Object {
        "addListener": [Function],
        "closeDrawer": [Function],
        "getState": [Function],
        "goBack": [Function],
        "isFocused": [Function],
        "navigate": [Function],
        "openDrawer": [Function],
        "pop": [Function],
        "popToTop": [Function],
        "push": [Function],
        "replace": [Function],
        "setParams": [Function],
        "toggleDrawer": [Function],
      }
    }
    route={
      Object {
        "name": "edgeLogin",
        "params": Object {
          "lobbyId": "AmNsSBDVeF2837",
        },
      }
    }
  />
</Context.Provider>
`;
